### DISCUSSION TITLE
Postgres: UPDATE with CTE is missing FROM clause

### DISCUSSION THREAD

**Discussion by wereii** [sqlalchemy/sqlalchemy#10407] [Usage Questions]
Hello, I am trying to create an atomic "insert then update" query by using CTE (WITH) postgres clause.

Currently I have this:

```py
class A(BaseModel):
 
  ...
  next_id: Mapped[int | None] = mapped_column(ForeignKey("a.id"))



insert_a_cte = (
      insert(A)
      .values(
         ...
      )
      .returning(A.id)
  ).cte("insert_a_cte")

update_query = (
      update(A)
      .values(next_id=insert_a_cte.c.id)
      .where(A.id == some_id)
      .add_cte(insert_a_cte)
  )
```

When running the `update_query` I get `[42P01] ERROR: missing FROM-clause entry for table "insert_a_cte"`. This is because the generated `UPDATE` is missing `FROM insert_a_cte`.  
I did find the pg `UPDATE .. FROM` mentioned in docs https://docs.sqlalchemy.org/en/20/tutorial/data_update.html#tutorial-update-from but that only mentions that the `FROM` clause is generated automatically when foreign tables are referenced (in `where`).

Is there a way to add that `FROM` manually in-some way?  Or am I doing something wierd/wrong here?

---
**OS:** Linux  
**Python:** 3.11.5
**SQLAlchemy:** 2.0.16
**Driver: asyncpg** 0.27.0

**Comment by zzzeek** [comment]
hi - 

can you please confirm this patch resolves:

```diff
diff --git a/lib/sqlalchemy/sql/dml.py b/lib/sqlalchemy/sql/dml.py
index 4047ba4133..d57f0401e2 100644
--- a/lib/sqlalchemy/sql/dml.py
+++ b/lib/sqlalchemy/sql/dml.py
@@ -211,7 +211,7 @@ class DMLState(CompileState):
         primary_table = all_tables[0]
         seen = {primary_table}
 
-        for crit in statement._where_criteria:
+        for crit in statement._where_criteria + statement._independent_ctes:
             for item in _from_objects(crit):
                 if not seen.intersection(item._cloned_set):
                     froms.append(item)

```

**Reply by zzzeek** [reply]
that patch is actually wrong, it will fix this immediate case but breaks others

**Reply by zzzeek** [reply]
new patch

```diff
diff --git a/lib/sqlalchemy/sql/dml.py b/lib/sqlalchemy/sql/dml.py
index d57f0401e2..13c4c9d779 100644
--- a/lib/sqlalchemy/sql/dml.py
+++ b/lib/sqlalchemy/sql/dml.py
@@ -211,7 +211,11 @@ class DMLState(CompileState):
         primary_table = all_tables[0]
         seen = {primary_table}
 
-        for crit in statement._where_criteria + statement._independent_ctes:
+        consider = statement._where_criteria
+        if self._dict_parameters and statement._values:
+            consider += tuple(statement._values.values())
+
+        for crit in consider:
             for item in _from_objects(crit):
                 if not seen.intersection(item._cloned_set):
                     froms.append(item)

```

**Reply by CaselIT** [reply]
IIRC we discussed supporting `select_from` (or similar) in an update in one of the past meeting, but different db do different things here, there isn't a single behavior, so nothing come from it.

That said that change makes sense to me

**Comment by zzzeek** [comment]
you can work aroudn this by including your CTE in the WHERE clause

```
        update_query = (
            update(a)
            .values(next_id=insert_a_cte.c.id)
            .where(a.c.id == 10)
            .where(insert_a_cte.c.id > 0)
            .add_cte(insert_a_cte)
        )
```
